Implement evaluator for inlining of StringLatin1.inflate

BradleyWood opened the pull request:
Implement evaluator for inlining of StringLatin1.inflate

Signed-off-by: BradleyWood <bradley.wood@ibm.com>

0xdaryl reviewed (changes requested):
This looks pretty good.  Good job figuring it all out.

I made a few comments throughout, but here are some of the bigger ones.

Please use the `Reg` suffix on all variables holding a `TR::Register` for consistency.

I would appreciate a few more descriptive comments in the main body of the implementation that describes the approach.  My experience with similar sequences in other areas of the code is that you may spend a bit of time re-grokking the code when you come back to it in the future.

If a node is commoned (i.e., its reference count is > 1), when you evaluate that node into a virtual register you can't make modifications to that register without affecting the other consumers of that node. The exception to this is if this is the last reference to a node (i.e., `node->getReferenceCount() == 1`)

In your case, once you evaluate the source offset, the destination offset, and length children into a register, you can't modify that register if there are future consumers of that node.

What we typically do in this situation is a "clobber evaluate" rather than an "evaluate" (e.g., `gprClobberEvaluate()` vs `evaluate()`).  A clobber evaluate is intended to be used in situations where the register will be clobbered by evaluator.  It will check the reference count before evaluating, and if > 1 then it will evaluate and then insert a register copy and hand you back the copied register.  You are free to manipulate the copy as you wish.

There are some opportunities to use smaller instructions, make better use of the addressing modes to make the code a bit cleaner, and move code to only the paths that it will be executed on.  I reworked parts of the implementation as an example. Please consider this as a guide only as I may not have fully grokked the full details of the implementation.

```
   headerOffsetConst = sizeof contiguous array header
   vectorLengthConst = 16

   lea   destBufferReg, [destBufferReg + 1*destOffsetReg + headerOffsetConst ]
   lea   srcOffsetReg, [srcOffsetReg + headerOffsetConst]
   lea   lengthReg, [lengthReg + 1*srcOffsetReg + -vectorLengthConst]

   cmp   srcOffsetReg, lengthReg  
   jg endLoop                  ; zero length or too small for vector length
                               ; jg instead of jge so that length=16 is done in the vector loop

   pxor  zeroReg, zeroReg

startLoop:

   movdqu xmmHighReg, [srcBufferReg + 1*srcOffsetReg]
   movdqu xmmLowReg, xmmHighReg

   punpckhbw xmmLowReg, zeroReg
   movdqu [destBufferReg + vectorLengthConst], xmmLowReg

   punpcklbw xmmHighReg, zeroReg
   movdqu [destBufferReg], xmmHighReg

   add   srcOffsetReg, vectorLengthConst
   add   destBufferReg, 2*vectorLengthConst

   cmp   srcOffsetReg, lengthReg
   jle startLoop

endLoop:

   add   lengthReg, vectorLengthConst

residueLoop:
   cmp   srcOffsetReg, lengthReg
   jge done

   ; Rather than copy up to 15 bytes individually, there may be an opportunity to use
   ; some subset of an 8 byte move + 4 byte move + 2 byte move + 1 byte move
   ; to move data more efficiently.  However, the cost of determining what to do at
   ; runtime and branching around the parts that don't apply may be prohibitive.
   ; If you haven't already, you might want to think about what that instruction
   ; sequence would look like and we can reason about its performance implications.

   movzx tempReg, [srcBufferReg + 1*srcOffsetReg]
   mov   [destBufferReg], tempReg
   inc   srcOffsetReg
   add   destOffsetReg, 2
   jmp residueLoop

done:

```

fjeremic commented on runtime/compiler/x/codegen/J9TreeEvaluator.cpp:
```diff
@@ -11584,6 +11589,101 @@ J9::X86::TreeEvaluator::directCallEvaluator(TR::Node *node, TR::CodeGenerator *c
+   generateLabelInstruction(JMP4, node, startLoop, cg);
+   generateLabelInstruction(LABEL, node, endLoop, cg);
+
+   // HANDLE REMAINDER (< 16 bytes)
+   TR::LabelSymbol *remStartLoop = generateLabelSymbol(cg);
+   TR::LabelSymbol *remEndLoop = generateLabelSymbol(cg);
+
+   remStartLoop->setStartInternalControlFlow();
+   remEndLoop->setEndInternalControlFlow();
+
+   generateRegImmInstruction(ADD4RegImm4, node, length, 16, cg);
+   generateRegRegInstruction(XOR4RegReg, node, gpr, gpr, cg);
+
+   generateLabelInstruction(LABEL, node, remStartLoop, cg);
+   generateRegRegInstruction(CMP4RegReg, node, srcOffset, length, cg);
+   generateLabelInstruction(JGE4, node, remEndLoop, cg);
+
+   generateRegMemInstruction(L1RegMem, node, gpr, generateX86MemoryReference(srcBuffer, srcOffset, 0, headerOffset, cg), cg);
+   generateMemRegInstruction(S2MemReg, node, generateX86MemoryReference(destBuffer, destOffset, 0, headerOffset, cg), gpr, cg);
+
+   // increase srcOffset by 1, destOffset by 2 to account for inflation
+   generateRegInstruction(INC4Reg, node, srcOffset, cg);
+   generateRegImmInstruction(ADD4RegImm4, node, destOffset, 2, cg);
+
+   generateLabelInstruction(JMP4, node, remStartLoop, cg);
+   generateLabelInstruction(LABEL, node, remEndLoop, cg);
```
You should benchmark this. I think an unrolled [Duff's device](https://en.wikipedia.org/wiki/Duff%27s_device) single branch implementation may be much faster than processing one byte at a time. We implemented this on Z and had a significant performance improvement for short lengths. You can touch base with @dchopra001 to see how it is implemented.

BradleyWood replied:
I believe I understand what you are suggesting, might it be more efficient to do the approach suggested by @0xdaryl, copying 8, then 4, then 2, then 1?

fjeremic replied:
Yes, it _should_ be better, but we'll have to benchmark it to find out for sure. On Z we split the code into 3 parts:

1. Process length >= 16
2. Process length >= 8 && <= 15
3. Process length < 8

We do the first two with vectors, and the last one using an unrolled Duff's device with a single branch. The entire evaluator I believe has only three branches.

Here is the benchmark we used:
https://github.com/fjeremic/jmh-benchmarks/blob/master/src/main/java/org/openj9/StringLatin1Benchmark.java

See the README in that repo on how to compile and run. It's fairly straightforward. We found on DayTrader7 lengths 0-4 account for ~80% of Strings that pass through this API, so accelerating short lengths to be as fast as possible will give us the biggest bang for the buck.

@dchopra001 could you work with @BradleyWood and sync up on all of this. Let's make sure we don't duplicate any work/investigation and help each other out.

fjeremic commented on runtime/compiler/x/codegen/J9TreeEvaluator.cpp:
```diff
@@ -11584,6 +11589,101 @@ J9::X86::TreeEvaluator::directCallEvaluator(TR::Node *node, TR::CodeGenerator *c
+   TR_ASSERT_FATAL(cg->getSupportsInlineStringLatin1Inflate(), "Inlining of StringLatin1.inflate not supported");
+   TR_ASSERT_FATAL(node->getNumChildren() == 5, "Wrong number of children in inlineStringLatin1Inflate");
+
+   intptr_t headerOffset = TR::Compiler->om.contiguousArrayHeaderSizeInBytes();
```
> How can I check for this scenario?

@dchopra001 can you sync up and show Bradley how this is done?

dchopra001 replied:
I missed this comment before. @BradleyWood you should be able to use the following API to do this: `TR::Compiler->om.canGenerateArraylets()`. 

Here's an example in J9CodeGenerator.cpp:
https://github.com/eclipse-openj9/openj9/blob/8e6702f56ea01de4df9ada5652820e0e75a73cc6/runtime/compiler/x/codegen/J9CodeGenerator.cpp#L115-L120

It'll be similar to the MMX check you had before in J9CodeGenerator.cpp.

BradleyWood commented:
I have changed the way the residual bytes (< 16) are copied to use the method suggested by @0xdaryl . The duffs device method did not yield good performance results. I have attached the benchmarks below. Performance is good, except for 0 length strings, which I suspect are not too common. I added a branch to compare the length of the string with zero, though it still could not match the non-intrinsic implementation.

![image](https://user-images.githubusercontent.com/14269423/118497990-90591f00-b6f3-11eb-8f94-34ef5001e19b.png)

Also,

```
   lea   destBufferReg, [destBufferReg + 1*destOffsetReg + headerOffsetConst ]
   lea   srcOffsetReg, [srcOffsetReg + headerOffsetConst]
   lea   lengthReg, [lengthReg + 1*srcOffsetReg + -vectorLengthConst]
```

This would require extra registers because they would need to be clobbered.

0xdaryl commented:
Thanks for the analysis.  Can you paste the generated code from your residue implementation using Duff's Device so I can see the instruction sequence please?

BradleyWood commented:
I have not had the chance to clean it up yet, so bear with me.

```
   TR::LabelSymbol *afterDuffDevice = generateLabelSymbol(cg);
   TR::LabelSymbol *duffDevice = generateLabelSymbol(cg);
   TR::LabelSymbol *afterCopy8 = generateLabelSymbol(cg);

   generateRegRegInstruction(SUB4RegReg, node, lengthReg, srcOffsetReg, deps, cg);

   generateRegImmInstruction(CMP4RegImm4, node, lengthReg, 8, deps, cg);
   generateLabelInstruction(JL1, node, afterDuffDevice, deps, cg);

   generateRegMemInstruction(MOVQRegMem, node, xmmLowReg, generateX86MemoryReference(srcBufferReg, srcOffsetReg, 0, headerOffsetConst, cg), deps, cg);
   generateRegRegInstruction(PUNPCKLBWRegReg, node, xmmLowReg, zeroReg, deps, cg);
   generateMemRegInstruction(MOVDQUMemReg, node, generateX86MemoryReference(destBufferReg, destOffsetReg, 0, headerOffsetConst, cg), xmmLowReg, deps, cg);
   generateRegImmInstruction(SUB4RegImm4, node, lengthReg, 8, deps, cg);
   generateRegImmInstruction(ADD4RegMem, node, srcOffsetReg, 8, deps, cg);
   generateRegImmInstruction(ADD4RegMem, node, srcOffsetReg, 16, deps, cg);

   generateLabelInstruction(LABEL, node, afterCopy8, deps, cg);

   generateRegRegImmInstruction(IMUL4RegRegImm4, node, scratchReg, lengthReg, -10, cg);
   generateRegImmInstruction(ADD4RegImm4, node, scratchReg, 80, deps, cg);

   generateRegMemInstruction(LEA4RegMem, node, lengthReg,  generateX86MemoryReference(duffDevice->getLabelSymbol(), cg), deps, cg);
   generateRegRegInstruction(ADD4RegReg, node, lengthReg, scratchReg, deps, cg);
   generateRegInstruction(JMP4, node, lengthReg, deps, cg);

   generateLabelInstruction(LABEL, node, duffDevice, deps, cg);

   for (int i = 0; i < 7; i++) {
      generateRegMemInstruction(MOVZXReg2Mem1, node, xmmLowReg, generateX86MemoryReference(srcBufferReg, srcOffsetReg, 0, headerOffsetConst + 7 - i, cg), cg);
      generateMemRegInstruction(S2MemReg, node, generateX86MemoryReference(destBufferReg, destOffsetReg, 0, headerOffsetConst + 2 * (7 - i), cg), xmmLowReg, cg);
   }

   generateLabelInstruction(LABEL, node, afterDuffDevice, deps, cg);
   afterDuffDevice->setEndInternalControlFlow();
```

fjeremic commented:
> I have attached the benchmarks below. Performance is good, except for 0 length strings, which I suspect are not too common.

@dchopra001 can you please paste the string length distribution chart you have from DayTrader7? I believe size 0 strings were fairly common even though I'd agree you would not expect that to be the case.

genie-openj9 commented:
Can one of the admins verify this patch?

dchopra001 commented:
Here's the distribution chart for Strings `length >=0 && length <= 20` (seen by Liberty in a DayTrader7 run):

![Screen Shot 2021-05-17 at 10 27 45 AM](https://user-images.githubusercontent.com/11260919/118505719-ad452080-b6fa-11eb-8ff9-2ee09f3baf07.png)

- ~96.5% of all Strings were observed to be within this length
- Strings of length 2 make up ~32% of all strings
- Strings of length 1 and 2 make up about 45-46% of all strings

fjeremic commented:
Right! It was _not_ 0 size. It was 1 & 2 that were by far the most common. My apologies I had that wrong. Given the results then we should perform better, though it would be best if we saw bigger improvements in the short string sizes.

fjeremic commented on runtime/compiler/x/codegen/J9TreeEvaluator.cpp:
```diff
@@ -11584,6 +11589,149 @@ J9::X86::TreeEvaluator::directCallEvaluator(TR::Node *node, TR::CodeGenerator *c
+   generateRegRegInstruction(CMP4RegReg, node, srcOffsetReg, scratchReg, deps, cg);
+   generateLabelInstruction(JG4, node, endLoopLabel, deps, cg);
+
+   // read 16 bytes into xmmr register
+   generateRegMemInstruction(MOVDQURegMem, node, xmmHighReg, generateX86MemoryReference(srcBufferReg, srcOffsetReg, 0, headerOffsetConst, cg), deps, cg);
+   // TODO; avoid this copy in the future by adding support for VPUNPCK instructions
+   generateRegRegInstruction(MOVDQURegReg, node, xmmLowReg, xmmHighReg, deps, cg);
+   // interleave the low order bytes with zeros and write to the destination buffer
+   generateRegRegInstruction(PUNPCKHBWRegReg, node, xmmLowReg, zeroReg, deps, cg);
+   generateMemRegInstruction(MOVDQUMemReg, node, generateX86MemoryReference(destBufferReg, destOffsetReg, 0, headerOffsetConst + vectorLengthConst, cg), xmmLowReg, deps, cg);
```
I see we're adding `deps` to all instructions in the ICF. This is not functionally wrong, but it will cause the RA to do more work than it needs to. Typically we only add the dependency to the merge label of the ICF.

BradleyWood commented:
Changed TR_X86OpCodes to TR::InstOpCode::Mnemonic

BradleyWood commented:
[Prefixed mnemonics with TR::InstOpCode](https://github.com/eclipse-openj9/openj9/compare/1834734bdcadae7ea99038e5ba45154db7bc8e16..2917266473984b0c018995582ef2af65563c4090)

fjeremic commented:
Jenkins test xlinux sanity jdk8,jdk11

fjeremic commented:
Jenkins test sanity xlinux jdk8,jdk11

fjeremic commented:
Jenkins test sanity xlinux jdk8,jdk11

fjeremic commented:
Jenkins test sanity xlinux jdk8,jdk11

0xdaryl commented on runtime/compiler/x/codegen/J9TreeEvaluator.cpp:
```diff
@@ -11924,6 +11930,157 @@ J9::X86::TreeEvaluator::directCallEvaluator(TR::Node *node, TR::CodeGenerator *c
+   deps->addPostCondition(destOffsetReg, TR::RealRegister::edx, cg);
+   deps->stopAddingConditions();
+
+   TR::LabelSymbol *afterDuffDevice = generateLabelSymbol(cg);
```
Call this `doneLabel`

0xdaryl commented on runtime/compiler/x/codegen/J9TreeEvaluator.cpp:
```diff
@@ -11924,6 +11930,157 @@ J9::X86::TreeEvaluator::directCallEvaluator(TR::Node *node, TR::CodeGenerator *c
+   deps->stopAddingConditions();
+
+   TR::LabelSymbol *afterDuffDevice = generateLabelSymbol(cg);
+   TR::LabelSymbol *duffDevice = generateLabelSymbol(cg);
```
Call this `copyResidueLabel`

0xdaryl commented on runtime/compiler/x/codegen/J9TreeEvaluator.cpp:
```diff
@@ -11924,6 +11930,157 @@ J9::X86::TreeEvaluator::directCallEvaluator(TR::Node *node, TR::CodeGenerator *c
+
+   TR::LabelSymbol *afterDuffDevice = generateLabelSymbol(cg);
+   TR::LabelSymbol *duffDevice = generateLabelSymbol(cg);
+   TR::LabelSymbol *afterCopy8 = generateLabelSymbol(cg);
```
Call this `afterCopy8Label`

0xdaryl commented on runtime/compiler/x/codegen/J9TreeEvaluator.cpp:
```diff
@@ -11924,6 +11930,157 @@ J9::X86::TreeEvaluator::directCallEvaluator(TR::Node *node, TR::CodeGenerator *c
+
+   bool is64bit = cg->comp()->target().is64Bit();
+   // calculate address to jump too
+   generateRegMemInstruction(TR::InstOpCode::LEARegMem(is64bit), node, scratch, generateX86MemoryReference(duffDevice, cg), cg);
```
This won't work without a relocation for AOT compiles (because the "duffDevice" address is only valid for this method, not if it is AOT loaded later).

Have a look at how the jump table in the tableEvaluator is implemented:

https://github.com/eclipse/omr/blob/d3053f922dd8f81535ea3396fbab3e8a51ac6e83/compiler/x/codegen/ControlFlowEvaluator.cpp#L468

BradleyWood replied:
Are you sure? The code generated is rip-relative,

`48 8d 05 0e 00 00 00 // lea    rax,[rip+0xe] `

Unless this is not guaranteed...

0xdaryl replied:
I wouldn't say "guaranteed", but I'd say it is extremely likely to produce RIP-relative code.  I'll double check the mem ref code.

Note that RIP-relative addressing is not available on 32-bit, so if this is something that will run on JDK8 then you'll need another solution (like the table evaluator)

BradleyWood replied:
This StringLatin1.inflate does not exist of JDK8

0xdaryl replied:
Maybe add a fatal assert at the beginning of this function then guarding against a 32-bit invocation.

BradleyWood replied:
See [force-push](https://github.com/eclipse-openj9/openj9/compare/f08e7ba55d1e2570b9d4401f8adca3e9edc9d0b4..42b99387f681bf1b6645f32d38baf1dea925d2a8)

0xdaryl commented:
Jenkins test sanity xlinux,win,osx jdk8,jdk17

0xdaryl commented:
@fjeremic : any chance you could approve your requested changes?

fjeremic commented on runtime/compiler/x/codegen/J9TreeEvaluator.cpp:
```diff
@@ -11924,6 +11930,159 @@ J9::X86::TreeEvaluator::directCallEvaluator(TR::Node *node, TR::CodeGenerator *c
+   // handle residual (< 8 bytes left) & jump to copy instructions based on the number of bytes left
+   // calculate how many bytes to skip based on length;
+
+   const int copy_instruction_size = 5  // size of MOVZXReg2Mem1
+                                    +4; // size of S2MemReg
+
+   // since copy_instruction_size could change depending on which registers are allocated to scratchReg, srcOffsetReg and destOffsetReg
+   // we reserve them to be eax, ecx, edx, respectively
+
+   generateRegRegImmInstruction(TR::InstOpCode::IMUL4RegRegImm4, node, lengthReg, lengthReg, -copy_instruction_size, cg);
+   generateRegImmInstruction(TR::InstOpCode::ADD4RegImm4, node, lengthReg, copy_instruction_size * 7, cg);
```
Not sure if there is a better way to do this. Seems fragile to changes but at least we have the comment explaining why it is so.

BradleyWood replied:
Yeah, there really is no way to know the size of the instructions without reserving the registers because the REX prefix might be used. Maybe later we could modify the register dependencies to request any register that does not require the rex prefix.

fjeremic reviewed (approved):
Gald to see this going in!
